pallets/liquidity-pools-gateway/queue/src/lib.rs

```diff
@@ -0,0 +1,261 @@
+		#[pallet::weight(T::WeightInfo::process_message())]
+		#[pallet::call_index(0)]
+		pub fn process_message(origin: OriginFor<T>, nonce: T::MessageNonce) -> DispatchResult {
+			ensure_signed(origin)?;
+
+			let message = MessageQueue::<T>::take(nonce).ok_or(Error::<T>::MessageNotFound)?;
+
+			match T::MessageProcessor::process(message.clone()) {
+				Ok(_) => {
+					Self::deposit_event(Event::<T>::MessageExecutionSuccess { nonce, message });
+
+					Ok(())
+				}
+				Err(e) => {
+					FailedMessageQueue::<T>::insert(nonce, (message.clone(), e.error));
+
+					Self::deposit_event(Event::<T>::MessageExecutionFailure {
+						nonce,
+						message,
+						error: e.error,
+					});
+
+					Ok(())
+				}
+			}
+		}
+
+		/// Convenience method for manually processing a failed message.
+		#[pallet::weight(T::WeightInfo::process_failed_message())]
+		#[pallet::call_index(1)]
+		pub fn process_failed_message(
+			origin: OriginFor<T>,
+			nonce: T::MessageNonce,
+		) -> DispatchResult {
+			ensure_signed(origin)?;
+
+			let (message, _) =
+				FailedMessageQueue::<T>::get(nonce).ok_or(Error::<T>::MessageNotFound)?;
+
+			match T::MessageProcessor::process(message.clone()) {
+				Ok(_) => {
+					Self::deposit_event(Event::<T>::MessageExecutionSuccess { nonce, message });
+
+					FailedMessageQueue::<T>::remove(nonce);
+
+					Ok(())
+				}
+				Err(e) => {
+					Self::deposit_event(Event::<T>::MessageExecutionFailure {
+						nonce,
+						message,
+						error: e.error,
+					});
+
+					Ok(())
+				}
+			}
+		}
+	}
+
+	impl<T: Config> Pallet<T> {
+		fn service_message_queue(max_weight: Weight) -> Weight {
+			let mut weight_used = Weight::zero();
+
+			let mut processed_entries = Vec::new();
+
+			for (nonce, message) in MessageQueue::<T>::iter() {
+				processed_entries.push(nonce);
+
+				let weight = match T::MessageProcessor::process(message.clone()) {
+					Ok(post_info) => {
+						Self::deposit_event(Event::MessageExecutionSuccess { nonce, message });
+
+						post_info
+							.actual_weight
+							.unwrap_or_default()
+							// Extra weight breakdown:
+							//
+							// 1 read for the message
+							// 1 write for the event
+							// 1 write for the message removal
+							.saturating_add(T::DbWeight::get().reads_writes(1, 2))
+					}
+					Err(e) => {
+						Self::deposit_event(Event::MessageExecutionFailure {
+							nonce,
+							message: message.clone(),
+							error: e.error,
+						});
+
+						FailedMessageQueue::<T>::insert(nonce, (message, e.error));
+
+						e.post_info
+							.actual_weight
+							.unwrap_or_default()
+							// Extra weight breakdown:
+							//
+							// 1 read for the message
+							// 1 write for the event
+							// 1 write for the failed message
+							// 1 write for the message removal
+							.saturating_add(T::DbWeight::get().reads_writes(1, 3))
+					}
+				};
+
+				weight_used = weight_used.saturating_add(weight);
+
+				if weight_used.all_gte(max_weight) {
```

Comment:
First time I've seen this! Whats the difference with `>=`?

Reply:
```
error[E0369]: binary operation `>=` cannot be applied to type `frame_support::weights::Weight
```

Reply:
Ohh, I though they implemented `Ord` on it. Thanks!